users/models.py

```diff
@@ -1,3 +1,385 @@
+
+        def validate_is_active(v):
+            return isinstance(v, bool)
+
+        def validate_is_staff(v):
+            return isinstance(v, bool)
+
+        def validate_email_field(v):
+            if not isinstance(v, str) or len(v) > 50:
+                return False
+            try:
+                validate_email(v)
+                return True
+            except ValidationError:
+                return False
+
+        allowed_fields = {
+            'first_name': validate_first_name,
+            'last_name': validate_last_name,
+            'user_phone': validate_user_phone,
+            'title': validate_title,
+            'role': validate_role,
+            'email': validate_email_field,
+            'is_active': validate_is_active,
+            'is_staff': validate_is_staff
+        }
+
+        for attr, validator in allowed_fields.items():
+            if attr in kwargs:
+                value = kwargs[attr]
+                if not validator(value):
+                    raise ValueError(f"Invalid value for field '{attr}': {value}")
+                setattr(self, attr, value)
+
+        if 'password' in kwargs:
+            password = kwargs['password']
+            if not isinstance(password, str) or len(password) < 8 or len(password) > 128:
+                raise ValueError("Password must be a string between 8 and 128 characters.")
+            self.set_password(password)
+
+        self.save()
+        return self
+
+    @classmethod
+    def delete_by_id(cls, user_id):
+        """
+        Deactivate a user by setting is_active to False.
+
+        Args:
+            user_id (int): The ID of the user to deactivate.
+
+        Returns:
+            bool: True if user was found and deactivated, else False.
+        """
+        user = cls.objects.filter(user_id=user_id, is_active=True).first()
+        if user:
+            user.is_active = False
+            user.save()
+            return True
+        return False
+
+    @classmethod
+    def get_all(cls):
+        """
+        Retrieve all users.
+
+        Returns:
+            QuerySet: QuerySet of all users.
+        """
+        return cls.objects.all()
+
+
+class UserRole(models.Model):
+    """
+    User role model.
+
+    Attributes:
+        user (ForeignKey): Related user instance.
+        role (str): Role name.
+        created_at (datetime): Creation timestamp.
+        updated_at (datetime): Last update timestamp.
+    """
+
+    user = models.ForeignKey(User, on_delete=models.CASCADE, related_name="roles")
+    role = models.CharField(max_length=20)
+    created_at = models.DateTimeField(default=timezone.now)
+    updated_at = models.DateTimeField(auto_now=True)
+
+    class Meta:
+        db_table = 'user_roles'
+
+    def __str__(self):
+        """
+        String representation of the user role.
+
+        Returns:
+            str: User email and role name.
+        """
+        return f"{self.user.email} - {self.role}"
```

Comment:
UserRole model allows multiple roles per user, but the User model has a role field as well.
Could you clarify the distinction and expected data flow? Is UserRole intended for history/audit or multi-role support?